### ISSUE TITLE
UP007: rule is still enabled on py39 for pyi

### ISSUE THREAD

**Issue by adifelice-godaddy** [astral-sh/ruff#7136]
## Overview
Possibly duplicating: https://github.com/astral-sh/ruff/issues/3011

According to documentation, the rule `UP007` should not be enabled if `target-version` is earlier than `py310`, but this does not apply to `pyi` or stub files.

## Details
`test.py`
```python3
from typing import Any, Optional


class Test:
    p: Optional[Any]
```

```
$ ruff check --select=UP007 --target-version=py39 test.py
$
```

`test.pyi`
```python3
from typing import Any, Optional


class Test:
    p: Optional[Any]
```

```shell
$ ruff check --select=UP007 --target-version=py39 test.pyi
test.pyi:5:8: UP007 [*] Use `X | Y` for type annotations
Found 1 error.
[*] 1 potentially fixable with the --fix option.
$
```

**Comment by charliermarsh** [comment]
Ah yeah, this is intentional, as `.pyi` files are treated as if `from __future__ import annotations` is enabled, and PEP 604-style unions are valid when used in such contexts. The other way to think of it is: `.pyi` files are only read by tools, and not at runtime, and so they can make use of language features that were added in later versions.

Here's another explanation with some references: https://github.com/astral-sh/ruff/issues/3851#issuecomment-1493203870.

**Comment by adifelice-godaddy** [comment]
We would prefer to align PROD and DEV python versions. Is there a way to change this behavior for `pyi`? Thanks.

**Comment by charliermarsh** [comment]
One option is to disable those rules for .pyi files via per-file-ignores. Something like:

```toml
[tool.ruff.per-file-ignores]
”*.pyi” = [“UP006”, “UP007”]
```

You could also just disable those rules entirely via the ignore setting if you don’t want them to be applied anywhere anyway.
